design/fd-support-kcp.md

```diff
@@ -0,0 +1,111 @@
+constitue two FDs. Labels placed on the BMHs, by the operator, can provide a
+means to distinguish between hosts belonging to different FDs.
+
+The proposal is for CAPM3 to choose hosts based on the FD selected by KCP.
+In other words, based on the value in `Machine.Spec.FailureDomain` CAPM3 will
+choose a BMH matching the label corresponding to that FD.
+
+## User Stories
+
+As an operator who has placed their baremetal infrastructure across different
+FDs, I would like CAPM3 to associate control-plane nodes with BMHs from the
+desired FD.
+
+## Goals
+
+* Support selection of BMHs based on the desired FD for the control plane node.
+
+## Non-Goals
+
+* Failure Domains support for worker nodes (i.e. MachineDeployments).
+
+## Proposal
+
+The operator labels the BMH resource with a standard FD label. For
+example, the following standard label could be used on the BMH:
+
+```diff
+  infrastructure.cluster.x-k8s.io/failure-domain=my-fd-1
+```
+
+Today, CAPM3 chooseHost() func associates a Metal3Machine with a specific BMH
+based on labels specified in `Metal3Machine.Spec.HostSelector.MatchLabels`.
+We can expand this capability. The HostSelector field is used to narrow down
+the set of available BMHs that meet the selection criteria. When FDs are
+being utilized (i.e. `Machine.Spec.FailureDomain` != nil and Machine is for a
+control-plane node), we insert the above label into HostSelector.MatchLabels.
+This should, in effect, narrow down the set of available BMHs to only those
+in the desired FD. Additional labels in the set could further narrow down the
+set of usable BMHs.
+
+Note that if no FD is specified in the `Machine.Spec.FailureDomain`, then the
+chooseHost() will default to it's current behavior.
+
+## FD labels on the Node object
+
+Kubernetes has the concept of [zones and regions.](https://kubernetes.io/docs/setup/best-practices/multiple-zones/)
+A zone can be considered an FD and 2+ zones fit within a logical grouping
+called a region. Nodes can be labeled with the well-known [zone label:](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone)
+
+```diff
+    topology.kubernetes.io/zone=my-fd-1
+    topology.kubernetes.io/zone=my-fd-2
+```
+
+For the purpose of scheduling pods (and placing pvcs), the zone labels can
+be utilized by the kubernetes scheduler to spread workloads across
+different FDs. While this proposal is only geared towards control-plane
+nodes, and zone labels are mostly applicable for worker nodes, we discuss
+some options here for operators who wish to have these labels applied to
+their control-plane nodes.
+
+### BMH Label Sync
+
+CAPM3 supports a [label sync mechanism](https://github.com/metal3-io/metal3-docs/blob/main/design/sync-labels-bmh-to-node.md)
+between the BMH and its corresponding Kubernetes Node. The sync controller
+could be configured to sync zone labels
+(e.g.`topology.kubernetes.io/zone=my-fd-1`) placed on the BMH to the
```

Comment:
would this work out of the box using label sync controller or I am not sure what you meant by could be configured?

Reply:
@furkatgofurov7 In this section on "FD labels on the Node Obj", I'm mentioning how information about what FD a Node belongs to can be bubbled up to the Node object. I'm doing this for completeness.

The label sync mechanism we have implemented can be configured to sync this zone label. The specifics on how to do this are in the linked proposal.